[FIX] resource, hr_work_entry_contract: split work intervals by type

Yoann-bary opened the pull request:
Steps to reproduce:
- Payroll > Configuration > Working Schedules
- Edit Monday morning to have shifts as follows -- 8:00-10:00, Overtime Hours type, 0.25 days
-- 10:00-12:00, Attendance type, 0.25 days
- Employees > Any with that schedule > Time Off
- Take a day off on a Monday (Full day)
- Time Off app > Management > Time Off
- Try to validate this time off

An error occurs when trying to read the morning's work entry type because both shifts have been merged (and thus the resulting shift has 2 entry types). This happens because of the use of Intervals in resource_calendar's '_attendance_intervals_batch' method.

When dealing with attendances we prefer using WorkIntervals which, unlike Intervals, keeps adjacent shifts separate. This is necessary to properly handle the entry type of leaves and attendances over contiguous periods.

This also removes the need for fdd9247cb86a01cc8c0903e69b85aae311e7c1ad, as the attendances have already been separated (Shifts cannot overlap any more than at their bounds) according to their type.

opw-4193334

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/182574.png)](https://mergebot.odoo.com/odoo/odoo/pull/182574)

nle-odoo reviewed (approved):
seems good to me

Bertrand2 reviewed (changes requested):
Just missing the update of the `.pot` file and a test, otherwise LGTM! :) Thanks

Yoann-bary commented:
I haven't touched any front-end text so I'm not sure what `.pot` you are referring to. I also reworked the test that was previously meant for _postprocess_attendance_intervals (Which I removed). If that's not enough I can add more but I think that might be somewhat redundant.

Bertrand2 commented:
> I haven't touched any front-end text so I'm not sure what `.pot` you are referring to. I also reworked the test that was previously meant for _postprocess_attendance_intervals (Which I removed). If that's not enough I can add more but I think that might be somewhat redundant.

Indeed, my bad! I somehow confused the test change as a code change, and the text inside it as a user text change :sweat_smile: Not enough coffee yet today it seems! LGTM

Bertrand2 commented:
robodoo r+

nle-odoo commented on addons/resource/models/resource_calendar.py:
```diff
@@ -20,6 +20,7 @@
 
 from odoo.tools import date_utils, float_utils
 from .utils import Intervals, float_to_time, make_aware, datetime_to_string, string_to_datetime
+from odoo.addons.hr_work_entry_contract.models.hr_work_intervals import WorkIntervals
```
hello


is this causing you an issue?

the cycle should not matter as far as I can tell (the python module should be cached, and gotten back from the cache, without importing the file a second time to follow the cycle)

as for hidden dependency to hr_work_entry_contract, do you mean that the python files will be loaded? then yes and it's not very clean but does it cause an issue to you?
